File: test/it_gas_x86_64_batch_0.py
import os
import subprocess as sub

import lib.assembly_map as asm
import lib.gen_code as gen

codeText = """
pro add_2_ints x0 x1
    x2 = add x0 x1
    ret x2

pro simple_inc x0
    x1 = add x0 1
    ret x1

pro simple_inc_test x0
    x1 = simple_inc x0
    ret x1

pro sum_lots_of_args x0 x1 x2 x3 x4 x5 x6 x7
    x10 = add x0 x1
    x11 = add x10 x2
    x12 = add x11 x3
    x13 = add x12 x4
    x14 = add x13 x5
    x15 = add x14 x6
    x16 = add x15 x7
    ret x16

pro swap_chars x0 x1
    x2 = [x0]
    x3 = [x1]
    [x0] = x3b
    [x1] = x2b
    ret 0

pro reverse_char_array x0 x1
    @begin
    jige x0 x1 @end
    x2 = [x0]
    x3 = [x1]
    [x0] = x3b
    [x1] = x2b
    x0 = add x0 1
    x1 = sub x1 1
    jump @begin
    @end
    ret 0

mem internal_ar {1 2 3}

pro sum_internal_ar x0
    x1 = internal_ar
    x2 = add x1 16
    x3 = 0
    @begin
    jig x1 x2 @end
    x4 = [x1]
    x3 = add x3 x4
    x1 = add x1 8
    jump @begin
    @end
    ret x3
"""

curDir = os.getcwd()
codeFile = 'it_gas_x86_64_batch_0'
codeFileLoc = './test/' + codeFile

def generateAssembly():
    asmap = asm.GasIntelX8664SystemVMap()
    code = gen.generateCodeLinesFromText(codeText, asmap)
    with open(f"{codeFileLoc}.asm", 'w') as file:
        for line in code:
            file.write(f"{line}\n")

def clearArea():
    for ext in ['.asm', '_c.o', '.o', '.out']:
        name = f"{codeFileLoc}{ext}"
        if os.path.exists(name):
            os.remove(name)

def runCodeTest():
    try:
        generateAssembly()
        sub.run(['as', '-msyntax=intel', '-mnaked-reg', '-o', f"{codeFileLoc}.o", f"{codeFileLoc}.asm"])
        sub.run(['gcc', '-c', '-o', f"{codeFileLoc}_c.o", f"{codeFileLoc}.c"])
        sub.run(['gcc', '-z', 'noexecstack', '-o', f"{codeFileLoc}.out", f"{codeFileLoc}_c.o", f"{codeFileLoc}.o"])
        sub.run([f"{codeFileLoc}.out"])
        clearArea()
    except Exception as ex:
        print('error during code generation')
        print(ex)

File: test/it_gas_x86_64_batch_0.c
#include <stdio.h>
#include <stdint.h>

extern int64_t add_2_ints(int64_t a, int64_t b);

void run_add_2_ints() {
	int64_t a = 1;
	int64_t b = 2;
    int64_t c = add_2_ints(a, b);
	printf("test 1:\n");
	printf("\t%lld + %lld = %lld\n", a, b, c);
}

extern int64_t simple_inc(int64_t a);
extern int64_t simple_inc_test(int64_t a);

void run_simple_inc_test() {
	int64_t a = 1;
    int64_t b = simple_inc_test(a);
	printf("test 2:\n");
	printf("\t%lld + 1 = %lld\n", a, b);
}

extern int64_t sum_lots_of_args(int64_t a, int64_t b, int64_t c,
	int64_t d, int64_t e, int64_t f, int64_t g, int64_t h);

void run_sum_lots_of_args() {
	int64_t s = sum_lots_of_args(1LL, 2LL, 3LL, 4LL,
		5LL, 6LL, 7LL, 8LL);
	printf("test 3:\n");
	printf("\t1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 = %lld\n", s);
}

extern int64_t swap_chars(char *a, char *b);

void run_swap_chars() {
	char a = 'a';
	char b = 'b';
	printf("test 4:\n");
	printf("\toriginal: %c %c\n", a, b);
	int64_t r = swap_chars(&a, &b);
	printf("\tswapped: %c %c\n", a, b);
}

extern int64_t reverse_char_array(char *begin, char *end);

void run_reverse_char_array() {
	char ar[6] = { 'a', 'b', 'c', 'd', 'e', '\0' };
	printf("test 5:\n");
	printf("\toriginal: %s\n", ar);
	int64_t r = reverse_char_array(ar, ar + 4);
	printf("\treversed: %s\n", ar);
}

extern int64_t sum_internal_ar(int64_t blank);

void run_sum_internal_ar() {
	int64_t s = sum_internal_ar(0LL);
	printf("test 6:\n");
	printf("\t1 + 2 + 3 = %lld\n", s);
}

int main(int argc, char **argv) {
	run_add_2_ints();
	run_simple_inc_test();
	run_sum_lots_of_args();
	run_swap_chars();
	run_reverse_char_array();
	run_sum_internal_ar();
	return 0;
}
